src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,343 @@
+                 maxVUBIncrement / 2, ProtocolSettings.Default.ValidatorsCount));
+            if (!CheckCommittee(engine)) throw new InvalidOperationException();
+            engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_MaxNotValidBeforeDelta))!.Set(value);
+        }
+
+        /// <summary>
+        /// GetNotaryNodes returns public keys of notary nodes.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <returns>Public keys of notary nodes.</returns>
+        private static ECPoint[] GetNotaryNodes(DataCache snapshot)
+        {
+            return RoleManagement.GetDesignatedByRole(snapshot, Role.P2PNotary, Ledger.CurrentIndex(snapshot) + 1);
+        }
+
+        /// <summary>
+        /// GetDepositFor returns Deposit for the specified account or nil in case if deposit
+        /// is not found in storage.
+        /// </summary>
+        /// <param name="snapshot"></param>
+        /// <param name="acc"></param>
+        /// <returns>Deposit for the specified account.</returns>
+        private Deposit? GetDepositFor(DataCache snapshot, UInt160 acc)
+        {
+            return snapshot.TryGet(CreateStorageKey(Prefix_Deposit, acc))?.GetInteroperable<Deposit>();
+        }
+
+        /// <summary>
+        /// PutDepositFor puts deposit on the balance of the specified account in the storage.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="acc">Account</param>
+        /// <param name="deposit">deposit</param>
+        private void PutDepositFor(ApplicationEngine engine, UInt160 acc, Deposit deposit)
+        {
+            // Don't need to seal because Deposit is a fixed-sized interoperable, hence always can be serialized.
+            var indeposit = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, acc), () => new StorageItem(deposit));
+            indeposit!.Value = new StorageItem(deposit).Value;
+        }
+
+        /// <summary>
+        /// RemoveDepositFor removes deposit from the storage.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="acc">Account</param>
+        private void RemoveDepositFor(DataCache snapshot, UInt160 acc)
+        {
+            snapshot.Delete(CreateStorageKey(Prefix_Deposit, acc));
+        }
+
+        /// <summary>
+        /// CalculateNotaryReward calculates the reward for a single notary node based on FEE's count and Notary nodes count.
+        /// </summary>
+        /// <param name="snapshot">DataCache</param>
+        /// <param name="nFees"></param>
+        /// <param name="notariesCount"></param>
+        /// <returns>result</returns>
+        private static long CalculateNotaryReward(IReadOnlyStore snapshot, long nFees, int notariesCount)
+        {
+            return nFees * Policy.GetAttributeFeeV1(snapshot, (byte)TransactionAttributeType.NotaryAssisted) / notariesCount;
```

Comment:
Are we sure that this matches:

```cs
if (notaryAssisted is not null)
                {
                    totalNetworkFee -= (notaryAssisted.NKeys + 1) * Policy.GetAttributeFeeV1(engine.SnapshotCache, (byte)notaryAssisted.Type);
                }
 ```
 
 There we multiple. Here we divide then mint, are the rounding correct?
 Maybe it would be safer to do the same calculation and then divide the sum correctly and precisely.
 
 ping @shargon

Reply:
The rounding is correct. We agree that there might be a situation when total notary reward is not divisible by `notariesCount`, then the reward per node will be rounded to lower integer value. The reminder will not be transferred to anyone, and we just live with it. So to us, it's not a problem since reminder is negligible.

If it's a problem, then we can always set NotaryAssisted attribute fee to be a multuple of `notariesCount`.

Reply:
> If it's a problem, then we can always set NotaryAssisted attribute fee to be a multuple of notariesCount.

It will be nice, every protection is good

Reply:
> It will be nice, every protection is good

It will be done by Committee via invocation of setAttributeFee.

Reply:
> > It will be nice, every protection is good
> 
> It will be done by Committee via invocation of setAttributeFee. So I don't think we need to update the code.

but they can do the opposite

Reply:
We still consider that this reminder is negligible and it doesn't worth to introduce dependency from RoleManagement into Policy contract to protect from possible committee misbehavior.

Reply:
Committee can do a lot of wrong things, protecting from bad committee is impossible in general. Realistically this is never a problem even if there are some datoshis burned.